flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/NestedPositionUtil.java

```diff
@@ -0,0 +1,209 @@
+                nullCollectionFlags.add(true);
+                nullValuesCount++;
+                offsets.add(++offset);
+                emptyCollectionFlags.add(false);
+            }
+        }
+        long[] offestsArray = offsets.toArray();
+        long[] length = calculateLengthByOffsets(emptyCollectionFlags.toArray(), offestsArray);
+        if (nullValuesCount == 0) {
+            return new CollectionPosition(Optional.empty(), offestsArray, length, valueCount);
+        }
+        return new CollectionPosition(
+                Optional.of(nullCollectionFlags.toArray()), offestsArray, length, valueCount);
+    }
+
+    public static boolean isOptionalFieldValueNull(int definitionLevel, int maxDefinitionLevel) {
+        return definitionLevel == maxDefinitionLevel - 1;
+    }
+
+    public static long[] calculateLengthByOffsets(
+            boolean[] collectionIsEmpty, long[] arrayOffsets) {
+        LongArrayList lengthList = new LongArrayList(arrayOffsets.length);
+        for (int i = 0; i < arrayOffsets.length - 1; i++) {
+            long offset = arrayOffsets[i];
+            long length = arrayOffsets[i + 1] - offset;
+            if (length < 0) {
+                throw new IllegalArgumentException(
+                        format(
+                                "Offset is not monotonically ascending. offsets[%s]=%s, offsets[%s]=%s",
+                                i, arrayOffsets[i], i + 1, arrayOffsets[i + 1]));
+            }
+            if (collectionIsEmpty[i]) {
+                length = 0;
+            }
+            lengthList.add(length);
+        }
+        return lengthList.toArray();
+    }
+
+    private static int getNextCollectionStartIndex(
+            int[] repetitionLevels, int maxRepetitionLevel, int elementIndex) {
+        do {
+            elementIndex++;
+        } while (hasMoreElements(repetitionLevels, elementIndex)
+                && !isCollectionBeginningMarker(
+                        repetitionLevels, maxRepetitionLevel, elementIndex));
+        return elementIndex;
+    }
+
+    /** This method is only called for non-empty collections. */
+    private static int getCollectionSize(
+            int[] repetitionLevels, int maxRepetitionLevel, int nextIndex) {
+        int size = 1;
+        while (hasMoreElements(repetitionLevels, nextIndex)
+                && !isCollectionBeginningMarker(repetitionLevels, maxRepetitionLevel, nextIndex)) {
+            // Collection elements cannot only be primitive, but also can have nested structure
+            // Counting only elements which belong to current collection, skipping inner elements of
+            // nested collections/structs
+            if (repetitionLevels[nextIndex] <= maxRepetitionLevel) {
+                size++;
+            }
+            nextIndex++;
+        }
+        return size;
+    }
+
+    private static boolean isCollectionBeginningMarker(
```

Comment:
revert this?
isNotCollectionBeginningMarker

Reply:
ok